network/src/topology.rs

```diff
@@ -169,9 +204,69 @@ impl KnownNetwork {
         self.connections.read().clone()
     }
 
-    /// Returns `true` if the known network contains any connections, `false` otherwise.
-    pub fn has_connections(&self) -> bool {
-        !self.connections.read().is_empty()
+    /// Returns a snapshot of all the nodes.
+    pub fn nodes(&self) -> HashMap<SocketAddr, u32> {
+        self.nodes.read().clone()
+    }
+
+    /// Constructs a set of nodes contained from the connection set.
+    pub fn nodes_from_connections(&self) -> HashSet<SocketAddr> {
+        let mut nodes: HashSet<SocketAddr> = HashSet::new();
+        for connection in self.connections().iter() {
+            // Using a hashset guarantees uniqueness.
+            nodes.insert(connection.source);
+            nodes.insert(connection.target);
+        }
+
+        nodes
+    }
+
+    /// Returns a map of the potential forks.
+    pub fn potential_forks(&self) -> HashMap<u32, Vec<SocketAddr>> {
+        use itertools::Itertools;
+
+        const HEIGHT_DELTA_TOLERANCE: u32 = 5;
+        const MIN_CLUSTER_SIZE: usize = 3;
+
+        let mut nodes: Vec<(SocketAddr, u32)> = self.nodes().into_iter().collect();
+        nodes.sort_unstable_by_key(|&(_, height)| height);
+
+        // Find the indexes at which the split the heights.
```

Comment:
I think you something in this comment :stuck_out_tongue:

Reply:
🤦